enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,283 @@
+---
+title: add-ImagePolicy-to-config.openshift.io-v1alpha1
+authors:
+  - "@QiWang19"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@saschagrunert, for API design, implementation details and graduation criteria"
+  - "@mrunalp"
+  - "@yuqi-zhang"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - "@mrunalp"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@mrunalp"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/OCPNODE-1628
+see-also:
+  - "/enhancements/this-other-neat-thing.md"
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# Add CRD ImagePolicy to config.openshift.io/v1alpha1
+
+## Summary
+
+This enhancement propose an ImagePolicy CRD to provide a way to set sigstore image signature verification configuration on Openshift.
+
+## Motivation
+
+### User Stories
+
+As an OpenShift user, I want to verify the container images signed using sigstore
+tools, so that I can utilize the increased security of my software supply chain.
+
+### Goals
+
+- The user can create an ImagePolicy instance specifying the images/repositories to be verified and their policy. The MCO will write the configuration for signature verification. Once this is done, CRI-O can verify the images/repositories.
+
+- MCO container runtime config controller watches ImagePolicy instance in different namespaces and merges the instances for each namespace into a single [containers-policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) in a predefined `/path/to/policies/\<NAMESPACE\>.json`.
+
+- CRI-O can verify Cosign signature signed images using configuration from ImagePolicy by using the namespace from the sandbox config on the `PullImage` RPC.
+
+### Non-Goals
+
+ImagePolicy CRD does not implements cosign with custom Fulcio and Rekor instance.
```

Comment:
What does this mean? That only the upstream `sigstore.dev` servers are accepted?

That effectively turns `sigstore.dev` into a single centralized identity provider, and a centralized record of all image signatures. Why would that be acceptable in an enterprise deployment?

Reply:
> What does this mean? That only the upstream sigstore.dev servers are accepted?

For the first iteration, yes. The option is to enable support for third party servers at some later feature stage.

@QiWang19 we may rephrase this sentence to make it a bit more clear.

Reply:
I’d rather advocate for enabling support for other ones immediately. It’s just two more fields to carry  (`fulcio.caData` and `rekorPublicKeyData`), structurally completely the same as the already proposed `keyData`.

That’s a completely trivial cost to implement the feature (admittedly there might be a noticeable cost to _test_ that it works with a real deployment of private servers), and it would send a strong message that we do expect a federated ecosystem to develop, instead of nudging early adopters to coalesce around a single resource (and excluding those who won’t).

---

Another thing to consider is that the underlying `policy.json` _requires_ the Fulcio and Rekor keys to be provided, so if the user won’t provide them via this CR, the MCO must obtain them … somehow.

I’m sure that code can be written, and it’s probably not too much code, but it’s a critical responsibility to take on. (And MCO will have to choose between including hard-coded public keys directly in the codebase and the shipped binary, and between fetching the public keys from some `sigstore.dev` TUF servers, which doesn’t work at all for completely disconnected systems.)

So, it seems to me to be much easier, in the first iteration, to _require_ the user to provide the two keys, and have MCO just move the data around. That is certain to be possible, and it immediately and without any extra work supports both disconnected clusters and private server instances; at the cost a bit of extra work when deploying. _Later_ some more native support for `sigstore.dev` itself could be added, where the users wouldn’t have to provide private keys.

Starting with a version that is simpler to implement but requires some extra user effort (mitigated by a 5-line shell script that can be updated if the keys change) seems to me to be a reasonable choice for a technical preview, or a first version, of a feature.

OTOH I could imagine a hypothetical where the only imperative were to provide an immediate and low-deployment-effort feature to work with `sigstore.dev`, and disconnected/private situations were not a concern at all; I’d just be very surprised. Is that the case here?

Reply:
> I’d rather advocate for enabling support for other ones immediately. It’s just two more fields to carry (fulcio.caData and rekorPublicKeyData), structurally completely the same as the already proposed keyData.

@QiWang19 we had a version where we included those fields, do we want to add them back?

Reply:
yes, I think we can add that back.

Reply:
I included the `fulcioCAData` and `rekorKeyData` in the structure.

Reply:
How should we update the field requirements?  If the `keyData` is set, we should require `oidcIssuer`, `subjectEmail`, `fulcioCAData` and `rekorKeyData` to be empty. If `keyData` is not set, we should require other values not to be empty?

Reply:
@mtrmac updated the API, could you take a look?

Reply:
The validation as described above makes sense to me, except that `keyData`+`rekorKeyData` is a valid combination.

The implemented validation expressions seem to not match precisely.

---

(Just to be sure, warning again: I’m not currently familiar with the k8s API evolution techniques, I’ve approved problematic designs in the past, and my opinion in that respect should carry no weight.)